### PR TITLE
Add attributes describing API stability.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Inspired by the conversation in #8015, adds two attributes which describe the stability of an API:

- `UnstableAttribute` indicates that we provide no API stability guarantees for the class between minor/patch versions. This is mostly applied to `*Impl` interfaces which describe the interface between core Avalonia and platform implementations, but also to some other APIs which are unlikely to be of use to client code, and are likely to change
- `NotClientImplementableAttribute` indicates that an interface is stable for consumption by a client, but should not be implemented outside of Avalonia itself (either because custom implementations are not supported, or because as members may be added to its API)

This can be used to suppress API Compat warnings for changes to unstable APIs, and later we may want to write an analyzer which produces warnings when an unstable type is used in client code.

I may have missed some types, or labeled some types incorrectly. Please let me know or push to this branch if so.

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020130-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Media/TextFormatting/ITextSource.cs [hunk 1]
```diff
@@ -1,8 +1,11 @@
-﻿namespace Avalonia.Media.TextFormatting
+﻿using Avalonia.Metadata;
+
+namespace Avalonia.Media.TextFormatting
 {
     /// <summary>
     /// Produces <see cref="TextRun"/> objects that are used by the <see cref="TextFormatter"/>.
     /// </summary>
+    [NotClientImplementable]
```
Comment (Gillibald) [inline src/Avalonia.Base/Media/TextFormatting/ITextSource.cs:8]: This is client implementable

Comment (Gillibald) [inline src/Avalonia.Base/Media/TextFormatting/ITextSource.cs:8]:
```suggestion
```

**Review by Gillibald — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020427-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020445-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020481-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**